base.py

```diff
@@ -1,20 +1,28 @@
 from abc import ABC, abstractmethod
 
-import joblib
 from joblib import delayed, effective_n_jobs, Parallel
 from typing import List, Union
 
 import numpy as np
 import pandas as pd
+from scipy.sparse import vstack, csr_array
```

Comment:
Can't we use `import scipy.sparse`? I worry that `vstack` in particular is the same as in Numpy

Reply:
Modified to import `scipy.sparse as sparse`